src/rt/ds/hashmap.h

```diff
@@ -110,144 +105,99 @@ namespace verona::rt
+    /**
+     * Return the probe index of the entry.
+     */
+    static uint8_t probe_index(uintptr_t key)
     {
-      Entry* old_set = set;
-      size_t old_size = get_size();
-      count = 0;
-
-      size_bits = marked > 3 ?
-        (uint8_t)snmalloc::bits::next_pow2_bits(marked << 1) :
-        INITIAL_SIZE_BITS;
-
-      size_t size = get_size();
-      assert(size > 0);
-
-      set = (Entry*)alloc->alloc<YesZero>(size * sizeof(Entry));
-
-      if (old_size != 0)
-      {
-        for (size_t index = 0; index < old_size; index++)
-        {
-          auto& entry = old_set[index];
-          auto& key = key_of(entry);
-
-          if ((key & MARK) != 0)
-          {
-            key = unmark_key(key);
-            insert(alloc, entry);
-          }
-          else if (key != 0)
-          {
-            delete_entry(entry);
-          }
-        }
-
-        alloc->dealloc(old_set, old_size * sizeof(Entry));
-      }
+      return (uint8_t)(key & PROBE_MASK);
     }
 
-    inline size_t get_dib(size_t size, size_t index, uintptr_t key)
+    /**
+     * Increase the allocation size by approximately 50%, rounded up to the
+     * nearest allocation size. The entries in the previous allocation will be
+     * reinserted.
+     */
+    void resize(Alloc* alloc)
     {
-      size_t dib = key & DIB_MAX;
+      auto prev = *this;
+
+      const auto alloc_size = capacity() * entry_alloc * 2;
+      slots = (Entry*)alloc->alloc<YesZero>(alloc_size);
+      capacity_shift = (uint8_t)bits::ctz(alloc_size / entry_alloc);
+      filled_slots = 0;
+      longest_probe = 0;
 
-      if (dib == DIB_MAX)
+      for (auto it = prev.begin(); it != prev.end(); ++it)
       {
-        // If we've encoded the maximum DIB, it could be an overflow.
-        // Recalculate the DIB.
-        size_t mask = size - 1;
-        dib = (index + size -
-               (verona::rt::bits::hash((void*)unmark_key(key)) & mask)) &
-          mask;
+        if constexpr (is_set)
+          insert(alloc, it.key());
+        else
+          insert(alloc, std::make_pair(it.key(), std::move(it.value())));
       }
-
-      return dib;
     }
 
+    /**
+     * Place an entry into the map at the given index, overwriting any existing
+     * entry. The probe bits of the key are set to `probe_len` and the
+     * `longest_probe` value is updated if `probe_len` is greater.
+     */
     template<typename E>
-    inline void set_entry(size_t index, E entry, size_t dib)
+    void place_entry(E entry, size_t index, uint8_t probe_len)
     {
-      // When entry is passed in, it may or may not have a mark bit, but it
-      // must have no dib. If the DIB is greater than the maximum DIB, encode
-      // it as the maximum DIB. We will recalculate the real DIB when we
-      // fetch it.
-      auto& key = key_of(entry);
-      key = key | (dib < DIB_MAX ? dib : DIB_MAX);
-      set[index] = std::forward<E>(entry);
+      slots[index] = std::forward<E>(entry);
```

Comment:
So you have allocated an array large enough for each object to be `entry_size` apart, but your indexing into the array does not use this.  So if you have a 3 word entry, the maths for the multiplication will be more complex than needed.  Can you make it so that everything is rounded up in the access.  Not sure what the nicest approach is:
* a wrapper type with alignment
* manually manage the array as a char* and do casting.  

I think the first would be easier to get right.